parlai/core/teachers.py

```diff
@@ -1944,6 +1945,223 @@ def shutdown(self):
+            raise ValueError(
+                'Chunk teacher should be used with streaming. '
+            )
+        if opt.get('pytorch_teacher_task') is not None:
+            raise ValueError(
+                'Chunk teacher is not compatible with pytorch '
+                'data teacher.'
+            )
+        if opt['numthreads'] > 1:
+            raise ValueError(
+                'Chunk teacher is not compatible with Hogwild.'
+            )
+
+        self.set_datasettings(opt['datatype'])
+        chunks = self.fold_chunks[:]
+
+        if (
+            shared is None
+            and self.is_train
+            and self.opt.get('distributed_world_size') is not None
+        ):
+            dws = int(self.opt['distributed_world_size'])
+            rank = int(self.opt['rank'])
+            self.fold_chunks = [c for c in self.fold_chunks if c % dws == rank]
+
+        if self.is_train:
+            random.Random().shuffle(chunks)
+
+        if shared is not None:
+            self.chunks = shared['chunks']
+            self.samples = shared['samples']
+            self.rng = shared['rng']
+            self.is_root_teacher = False
+        else:
+            self.is_root_teacher = True
+            self.samples = queue.Queue(maxsize=self.buffersize)
+            self.chunks = queue.Queue()
+            if self.is_train:
+                self.rng = random.Random()
+            else:
+                self.rng = random.Random(4)
+            self._enqueue_chunks()
+            # launch queue loader on the main thread
+            self._enqueue_request()
+
+        self.episode_done = True
+
+    def _get_data_folder(self):
+        if not self.opt.get('datafile'):
+            raise RuntimeError(
+                'Must specify datafile or override this function '
+                'to return the data folder.'
+            )
+
+        return self.opt['datafile']
+
+    @abstractmethod
+    def _get_num_samples(self, datatype) -> int:
+        """
+        [Abstract] Return the number of samples given the datatype.
+        """
+        pass
+
+    @abstractmethod
+    def _get_fold_chunks(self, datatype) -> List[int]:   # type: ignore
+        """
+        [Abstract] Return a list of chunk IDs (integer).
+
+        Given the datatype (train/test/valid), return the list of
+        chunk IDs that correspond to that split.
+        """
+        pass
+
+    def set_datasettings(self, datatype):
+        self.folder = self._get_data_folder()
+        self.num_samples = self._get_num_samples(datatype)
+        self.fold_chunks = self._get_fold_chunks(datatype)
+
+        self.is_train = 'train' in datatype and 'evalmode' not in datatype
+
+    def share(self):
+        shared = super().share()
+        shared['samples'] = self.samples
+        shared['chunks'] = self.chunks
+        shared['rng'] = self.rng
+        return shared
+
+    def _setup_data(self, datatype):
+        pass
+
+    def num_episodes(self):
```

Comment:
this isn't necessarily true. It was in my [redacted] teacher and might be in yours, but not in the general case.